runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ConstructionService.java

```diff
@@ -0,0 +1,214 @@
+    LOG.info(
+        "Expanding '{}' with URN '{}'",
+        request.getTransform().getUniqueName(),
+        request.getTransform().getSpec().getUrn());
+    LOG.debug("Full transform: {}", request.getTransform());
+    Set<String> existingTransformIds = request.getComponents().getTransformsMap().keySet();
+    Pipeline pipeline = Pipeline.create();
+    RehydratedComponents rehydratedComponents =
+        RehydratedComponents.forComponents(request.getComponents()).withPipeline(pipeline);
+
+    Map<String, PCollection<?>> inputs =
+        request
+            .getTransform()
+            .getInputsMap()
+            .entrySet()
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey,
+                    input -> {
+                      try {
+                        return rehydratedComponents.getPCollection(input.getValue());
+                      } catch (IOException exn) {
+                        throw new RuntimeException(exn);
+                      }
+                    }));
+    if (!registeredTransforms.containsKey(request.getTransform().getSpec().getUrn())) {
+      throw new UnsupportedOperationException(
+          "Unknown urn: " + request.getTransform().getSpec().getUrn());
+    }
+    registeredTransforms
+        .get(request.getTransform().getSpec().getUrn())
+        .apply(
+            pipeline,
+            request.getTransform().getUniqueName(),
+            request.getTransform().getSpec(),
+            inputs);
+
+    // Needed to find which transform was new...
+    RunnerApi.Pipeline pipelineProto =
+        PipelineTranslation.toProto(
+            pipeline,
+            rehydratedComponents.getSdkComponents().withNamespace(request.getNamespace()));
+    String expandedTransformId =
+        Iterables.getOnlyElement(
+            pipelineProto
+                .getRootTransformIdsList()
+                .stream()
+                .filter(id -> !existingTransformIds.contains(id))
+                .collect(Collectors.toList()));
+    RunnerApi.Components components = pipelineProto.getComponents();
+    LOG.debug("Expanded to {}", components.getTransformsOrThrow(expandedTransformId));
+
+    return ConstructionApi.ConstructionResponse.newBuilder()
+        .setComponents(components)
+        .setTransformId(expandedTransformId)
+        .build();
+  }
+
+  @Override
+  public void construct(
+      ConstructionApi.ConstructionRequest request,
+      StreamObserver<ConstructionApi.ConstructionResponse> responseObserver) {
+    try {
+      responseObserver.onNext(construct(request));
+      responseObserver.onCompleted();
+    } catch (RuntimeException exn) {
+      responseObserver.onError(exn);
+      throw exn;
+    }
+  }
+
+  public static void main(String[] args) throws Exception {
+    int port = Integer.parseInt(args[0]);
+    System.out.println("Starting construction service at localhost:" + port);
```

Comment:
When will we be starting this service ? 

Is this a part of SDK job-submission or something that has a different life-cycle ?

Reply:
It has a different lifecycle. An SDK may know about certain transforms and start up services on its own to expand them (in which case it'd probably be part of some WrapperTransform's expand method).